README.md
[![stable: 0.6.6](https://img.shields.io/badge/stable-0.6.6-ED2E26.svg?style=flat-square)](https://github.com/rodekruis/caladrius)
[![F.A.C.T.: 42](https://img.shields.io/badge/F\.A\.C\.T\.-42-291AE0.svg?style=flat-square)](https://rodekruis.sharepoint.com/sites/510-Team/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={FD66FFCB-C34C-433E-9706-F672A8EFAB3D})
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![code style: black](https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square)](https://github.com/psf/black)

# [Caladrius](https://en.wikipedia.org/wiki/Caladrius) - Assessing Building Damage caused by Natural Disasters using Satellite Images

## Created by: Artificial Incompetence for the Red Cross #1 Challenge in the 2018 Hackathon for Peace, Justice and Security

*Note: Parts of this project are not made public for privacy and operational reasons. If you would like to access any restricted content, please send an email to grahman@rodekruis.nl with the relevant details (content url, purpose of request, et cetera).*

## Documentation
1. [Presentation 2020](https://rodekruis.sharepoint.com/sites/510-Team/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={0DCCBEB8-91CC-4021-A0DF-CC5239CC5EBD})
2. [Project Specification Document](https://rodekruis.sharepoint.com/sites/510-Team/Gedeelde%20%20documenten/%5BPRJ%5D%20Automated%20Damage%20Assessment/DOCUMENTATION/caladrius.pdf)
3. [Presentation 2019](https://rodekruis.sharepoint.com/sites/510-Team/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={CD3544E2-69A7-4389-AEA0-7349DCA2A6DD})
4. [Performance Spreadsheet](https://rodekruis.sharepoint.com/sites/510-Team/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={5F134334-1E86-448D-BE78-BE76C1C7ED22})

## Network Architecture

The network architecture is a pseudo-[siamese network](http://papers.nips.cc/paper/769-signature-verification-using-a-siamese-time-delay-neural-network) with two [ImageNet](https://ieeexplore.ieee.org/abstract/document/5206848)
pre-trained [Inception_v3](http://openaccess.thecvf.com/content_cvpr_2016/html/Szegedy_Rethinking_the_Inception_CVPR_2016_paper.html) models.

## Using Docker

Install [Docker](https://www.docker.com/get-started).

Download the
[Caladrius Docker Image](https://hub.docker.com/r/gulfaraz/caladrius) using,

```
docker pull gulfaraz/caladrius
```

Create a [data](#dataset) folder in your local machine.

Create a docker container using,

```
docker run --name caladrius -dit -v <path/to/data>:/workspace/data -p 5000:5000 gulfaraz/caladrius
```

Access the container using,

```
docker exec -it caladrius bash
```

## Manual Setup

#### Requirements:

-   [Python 3.6.5](https://www.python.org/downloads/)
-   [Anaconda or Miniconda 2019.07](https://www.anaconda.com/distribution/#download-section)
-   [NodeJS v10](https://nodejs.org/en/download/)
-   Run the following script,

```bash
./caladrius_install.sh
```

## Dataset - Sint Maarten 2017 Hurricane Irma

##### 1. Download Raw Dataset:

The Sint Maarten 2017 dataset can be downloaded from
[here](https://rodekruis.sharepoint.com/sites/510-Team/Gedeelde%20%20documenten/%5BPRJ%5D%20Automated%20Damage%20Assessment/DATASET/Sint-Maarten-2017/rc.tgz "RC Challenge 1 Raw Dataset").

##### 2. Extract Raw Dataset:

To extract the contents to the `data` folder execute,

```
tar -xvzf rc.tgz
```

##### 3. Create Training Dataset:

Transform the raw dataset to a training dataset using,

```
python caladrius/dataset/sint_maarten_2017.py --version 1.0.0 --create-image-stamps --query-address-api --address-api openmapquest --address-api-key <ADDRESS_API_KEY> --create-report-info-file
```

The above command will create the dataset as per the
[specifications](DATASET.md).

##### Configuration:

`sint_maarten_2017.py` accepts the command line arguments described below,

```
usage: sint_maarten_2017.py [-h] --version VERSION [--create-image-stamps]
                            [--query-address-api] [--address-api ADDRESS_API]
                            [--address-api-key ADDRESS_API_KEY]
                            [--create-report-info-file]
                            [--label-type label_type]

optional arguments:
  -h, --help            show this help message and exit
  --version VERSION     set a version number to identify dataset (default:
                        None)
  --create-image-stamps
                        For each building shape, creates a before and after
                        image stamp for the learning model, and places them in
                        the approriate directory (train, validation, or test)
                        (default: False)
  --query-address-api   For each building centroid, preforms a reverse geocode
                        query and stores the address in a cache file (default:
                        False)
  --address-api ADDRESS_API
                        Which API to use for the address query (default:
                        openmapquest)
  --address-api-key ADDRESS_API_KEY
                        Some APIs (like OpenMapQuest) require an API key
                        (default: None)
  --create-report-info-file
                        Creates a geojson file that contains the locations and
                        shapes of the buildings, their respective
                        administrative regions and addresses (if --query-
                        address-api has been run) (default: False)
  --label-type label_type
                        Sets whether the damage label should be produced on a
                        continuous scale or in classes. (default: regression)
```

## Interface

From the `caladrius/interface` directory execute,

```
npm start
```

The interface should be accessible at `http://localhost:5000`.

## Model

##### Training:

```
python caladrius/run.py --run-name caladrius_2019
```

##### Testing:

```
python caladrius/run.py --run-name caladrius_2019 --test
```

[Click here to download the trained model.](https://rodekruis.sharepoint.com/sites/510-Team/Gedeelde%20%20documenten/%5BPRJ%5D%20Automated%20Damage%20Assessment/MODEL/Sint-Maarten-2017/Sint-Maarten-2017v0.4.tgz)

## Configuration

`run.py` accepts the command line arguments described below,

```
usage: run.py [-h] [--checkpoint-path CHECKPOINT_PATH] [--data-path DATA_PATH]
              [--run-name RUN_NAME] [--log-step LOG_STEP]
              [--number-of-workers NUMBER_OF_WORKERS]
              [--model-type {quasi-siamese,random,average}] [--disable-cuda]
              [--cuda-device CUDA_DEVICE] [--torch-seed TORCH_SEED]
              [--input-size INPUT_SIZE] [--number-of-epochs NUMBER_OF_EPOCHS]
              [--batch-size BATCH_SIZE] [--learning-rate LEARNING_RATE]
              [--test] [--max-data-points MAX_DATA_POINTS]
              [--train-accuracy-threshold TRAIN_ACCURACY_THRESHOLD]
              [--test-accuracy-threshold TEST_ACCURACY_THRESHOLD]
              [--output-type {regression,classification}]

optional arguments:
  -h, --help            show this help message and exit
  --checkpoint-path CHECKPOINT_PATH
                        output path (default: ./runs)
  --data-path DATA_PATH
                        data path (default: ./data/Sint-Maarten-2017)
  --run-name RUN_NAME   name to identify execution (default: <timestamp>)
  --log-step LOG_STEP   batch step size for logging information (default: 100)
  --number-of-workers NUMBER_OF_WORKERS
                        number of threads used by data loader (default: 8)
  --model-type {quasi-siamese,random,average}
                        type of model (default: quasi-siamese)
  --disable-cuda        disable the use of CUDA (default: False)
  --cuda-device CUDA_DEVICE
                        specify which GPU to use (default: 0)
  --torch-seed TORCH_SEED
                        set a torch seed (default: 42)
  --input-size INPUT_SIZE
                        extent of input layer in the network (default: 32)
  --number-of-epochs NUMBER_OF_EPOCHS
                        number of epochs for training (default: 100)
  --batch-size BATCH_SIZE
                        batch size for training (default: 32)
  --learning-rate LEARNING_RATE
                        learning rate for training (default: 0.001)
  --test                test the model on the test set instead of training
                        (default: False)
  --max-data-points MAX_DATA_POINTS
                        limit the total number of data points used, for
                        debugging on GPU-less laptops (default: None)
  --train-accuracy-threshold TRAIN_ACCURACY_THRESHOLD
                        window size to calculate regression accuracy (default:
                        0.1)
  --test-accuracy-threshold TEST_ACCURACY_THRESHOLD
                        window size to calculate regression accuracy (default:
                        0.3)
  --output-type {regression,classification}
                        choose if want regression or classification model
                        (default: regression)
```
---
## Development

### How to setup code for developement?

After cloning the repo, run `pre-commit install` to enable format checking when committing changes.

### How to manage versions?

When making changes, increment version number in [VERSION](VERSION), [package.json](caladrius/interface/package.json), the badge in [README.md](README.md) and [package.json](caladrius/interface/client/package.json) according to [PEP 440](https://www.python.org/dev/peps/pep-0440/) and update [CHANGES.md](CHANGES.md).

### How to build Docker image?

To build and tag the Docker image with [VERSION](VERSION),

```bash
make build_production
```

For development and tagging with the latest commit version,

```bash
make build_fast
```


CHANGES.md
0.1.0 (2021-03-04)
------------------
- Initial version


caladrius/utils.py
import os
import sys
import time
import argparse
import pickle
import logging
import re
import json

import torch

NEURAL_MODELS = ["inception", "light", "after", "shared", "vgg", "attentive"]
STATISTICAL_MODELS = ["average", "random"]

# logging

logging.getLogger("Fiona").setLevel(logging.ERROR)
logging.getLogger("fiona.collection").setLevel(logging.ERROR)
logging.getLogger("rasterio").setLevel(logging.ERROR)
logging.getLogger("PIL.PngImagePlugin").setLevel(logging.ERROR)


class dotdict(dict):
    """
    a dictionary that supports dot notation
    as well as dictionary access notation
    usage: d = dotdict() or d = dotdict({'val1':'first'})
    set attributes: d.val2 = 'second' or d['val2'] = 'second'
    get attributes: d.val2 or d['val2']
    """

    __getattr__ = dict.__getitem__
    __setattr__ = dict.__setitem__
    __delattr__ = dict.__delitem__

    def __init__(self, dct):
        super().__init__()
        for key, value in dct.items():
            if hasattr(value, "keys"):
                value = dotdict(value)
            self[key] = value


def make_directory(directoryPath):
    if not os.path.isdir(directoryPath):
        os.makedirs(directoryPath)
    return directoryPath


def save_obj(obj, path):
    with open(path, "wb") as f:
        pickle.dump(obj, f, pickle.HIGHEST_PROTOCOL)


def load_obj(path):
    with open(path, "rb") as f:
        return pickle.load(f)


def readable_float(number):
    return round(float(number), 4)


def dynamic_report_key(label, prefix, condition):
    return "{}{}".format((prefix + "_model_") if condition else "", label)


def run_name_type(run_name):
    run_name = str(run_name)
    pattern = re.compile(r"^[a-zA-Z0-9_\.]{3,30}$")
    if not pattern.match(run_name):
        raise argparse.ArgumentTypeError(
            "Run name can contain only "
            + "alphanumeric, underscore (_) and dot (.) characters. "
            + "Must be at least 3 characters and at most 30 characters long."
        )
    return run_name


def configuration():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    # General arguments
    parser.add_argument(
        "--checkpoint-path",
        type=str,
        default=os.path.join(".", "runs"),
        help="output path",
    )
    parser.add_argument(
        "--data-path",
        type=str,
        default=os.path.join(".", "data", "Sint-Maarten-2017"),
        help="data path",
    )
    parser.add_argument(
        "--model-path",
        type=str,
        default=os.path.join(".", "caladrius-models", "best_model_wts.pkl"),
        help="data path",
    )
    parser.add_argument(
        "--label-file", type=str, default="labels.txt", help="filename of labels",
    )
    parser.add_argument(
        "--run-name",
        type=run_name_type,
        default="{:.0f}".format(time.time()),
        help="name to identify execution",
    )
    parser.add_argument(
        "--log-step",
        type=int,
        default=10,
        help="batch step size for logging information",
    )
    parser.add_argument(
        "--number-of-workers",
        type=int,
        default=8,
        help="number of threads used by data loader",
    )
    parser.add_argument(
        "--model-type",
        type=str,
        default=NEURAL_MODELS[0],
        choices=NEURAL_MODELS + STATISTICAL_MODELS,
        help="type of model",
    )

    parser.add_argument(
        "--disable-cuda", action="store_true", help="disable the use of CUDA"
    )
    parser.add_argument(
        "--cuda-device", type=int, default=0, help="specify which GPU to use"
    )
    parser.add_argument("--torch-seed", type=int, help="set a torch seed", default=42)

    parser.add_argument(
        "--input-size",
        type=int,
        default=32,
        help="extent of input layer in the network",
    )
    parser.add_argument(
        "--number-of-epochs",
        type=int,
        default=100,
        help="number of epochs for training",
    )
    parser.add_argument(
        "--batch-size", type=int, default=32, help="batch size for training"
    )
    parser.add_argument(
        "--learning-rate", type=float, default=0.001, help="learning rate for training"
    )

    parser.add_argument(
        "--test",
        action="store_true",
        default=False,
        help="test the model on the test set instead of training",
    )
    parser.add_argument(
        "--inference",
        action="store_true",
        default=False,
        help="test and use the model on the inference set instead of training",
    )
    parser.add_argument(
        "--max-data-points",
        default=None,
        type=int,
        help="limit the total number of data points used, for debugging on GPU-less laptops",
    )
    parser.add_argument(
        "--output-type",
        type=str,
        default="regression",
        choices=["regression", "classification"],
        help="choose if want regression or classification model",
    )

    parser.add_argument(
        "--number-classes", type=int, default=4,
    )

    parser.add_argument(
        "--selection-metric",
        type=str,
        default="f1_macro",
        choices=[
            "recall_micro",
            "recall_macro",
            "precision_macro",
            "f1_macro",
            "recall_weighted",
            "precision_weighted",
            "f1_weighted",
        ],
        help="choose metric to use for tracking best model",
    )

    parser.add_argument(
        "--test-epoch",
        default=False,
        action="store_true",
        help="If true, run model on test set every epoch. For research purposes.",
    )

    parser.add_argument(
        "--sample-data",
        default=False,
        action="store_true",
        help="If true, resample data such that classes are balanced. For research purposes.",
    )

    parser.add_argument(
        "--weighted-loss",
        default=False,
        action="store_true",
        help="If True, the loss will be weighted according to the amount of data per damage category",
    )

    parser.add_argument(
        "--classification-loss-type",
        default="cross-entropy",
        choices=["cross-entropy", "f1"]
    )

    parser.add_argument(
        "--freeze",
        default=False,
        action="store_true",
        help="If true, Inception part will not be retrained.",
    )

    parser.add_argument(
        "--no-augment",
        default=False,
        action="store_true",
        help="If False, no augmentations will be applied to the data.",
    )

    parser.add_argument(
        "--augment-type",
        type=str,
        default="original",
        choices=["original", "paper", "equalization"],
        help="choose which data augmentation steps should be applied",
    )

    parser.add_argument(
        "--save-all",
        default=False,
        action="store_true",
        help="If True, whole model will be saved not only state dict. Only for testing purposes",
    )

    parser.add_argument(
        "--probability",
        default=False,
        action="store_true",
        help="If True, probabilistic predictions will be given",
    )

    args = parser.parse_args()

    arg_vars = vars(args)
    arg_vars["model_name"] = arg_vars["run_name"]

    if args.torch_seed is not None:
        torch.manual_seed(arg_vars["torch_seed"])
    else:
        arg_vars["torch_seed"] = torch.initial_seed()

    if args.max_data_points is not None:
        arg_vars["run_name"] = "{}_max_data_points_{}".format(
            arg_vars["run_name"], arg_vars["max_data_points"]
        )

    arg_vars[
        "model_directory"
    ] = "{}-input_size_{}-learning_rate_{}-batch_size_{}".format(
        arg_vars["run_name"],
        arg_vars["input_size"],
        arg_vars["learning_rate"],
        arg_vars["batch_size"],
    )

    arg_vars["checkpoint_path"] = make_directory(
        os.path.join(arg_vars["checkpoint_path"], arg_vars["model_directory"])
    )
    arg_vars["trained_model_path"] = os.path.join(
        arg_vars["checkpoint_path"], "best_model_wts.pkl"
    )
    arg_vars["prediction_path"] = make_directory(
        os.path.join(arg_vars["checkpoint_path"], "predictions")
    )
    arg_vars["run_report_path"] = os.path.join(
        arg_vars["checkpoint_path"], "run_report.json"
    )
    arg_vars["statistical_model"] = arg_vars["model_type"] in STATISTICAL_MODELS
    arg_vars["neural_model"] = arg_vars["model_type"] in NEURAL_MODELS

    if torch.cuda.is_available() and not arg_vars["disable_cuda"]:
        arg_vars["device"] = torch.device("cuda:{}".format(arg_vars["cuda_device"]))
    else:
        arg_vars["device"] = torch.device("cpu")

    return args


def load_run_report(run_report_path):
    run_report_json = dotdict({})
    if os.path.exists(run_report_path):
        with open(run_report_path, "r") as run_report_file:
            run_report_json = json.load(run_report_file)
            run_report_json = dotdict(run_report_json)
            run_report_json.device = torch.device(run_report_json.device)
    return run_report_json


def save_run_report(run_report_json):
    run_report_json.device = str(run_report_json.device)
    with open(run_report_json.run_report_path, "w") as run_report_file:
        json.dump(run_report_json, run_report_file, indent=4)


def attach_exception_hook(logger):
    def exception_logger(exceptionType, exceptionValue, exceptionTraceback):
        logger.error(
            "Uncaught Exception",
            exc_info=(exceptionType, exceptionValue, exceptionTraceback),
        )

    return exception_logger


def create_logger(module_name):
    args = configuration()

    debug_filehandler = logging.FileHandler(
        os.path.join(args.checkpoint_path, "run_debug.log")
    )
    info_filehandler = logging.FileHandler(
        os.path.join(args.checkpoint_path, "run_info.log")
    )

    formatter = logging.Formatter("%(asctime)s %(name)s %(levelname)s %(message)s")
    debug_filehandler.setFormatter(formatter)
    info_filehandler.setFormatter(formatter)

    debug_filehandler.setLevel(logging.DEBUG)
    info_filehandler.setLevel(logging.INFO)

    streamhandler = logging.StreamHandler(sys.stdout)
    streamhandler.setFormatter(formatter)
    streamhandler.setLevel(logging.DEBUG)

    logger = logging.getLogger(module_name)

    logger.addHandler(debug_filehandler)
    logger.addHandler(info_filehandler)
    logger.addHandler(streamhandler)

    logger.setLevel(logging.DEBUG)

    return logger
